### PR TITLE
Drop anemone and use Spidr for Repo discovery

### PR DESCRIPTION
**Pull request by sjha4**
To Do:

- [x] Support basic auth
- [x] Support proxy

#### What are the testing steps for this pull request?
bundle install
Go to Content > Product > Repo discovery
Run repo discovery.

### CODE REVIEW

**Comment by sjha4** [comment]
@evgeni : Thoughts on [spidr gem](https://github.com/postmodern/spidr/) as replacement for anemone?

**Comment by evgeni** [comment]
Doesn't look crazy? Only dep is nokogiri, which we have anyway, tested on modern rubies. Why not.

File: app/lib/katello/repo_discovery.rb
Comment (jeremylenz) [inline app/lib/katello/repo_discovery.rb:206]: These should probably be `Rails.logger.debug`s, no?

Comment (sjha4) [reply app/lib/katello/repo_discovery.rb:206]: Ya..I will clean this up and move it to ready for review.. 😻

**Comment by sjha4** [comment]
I am seeing significant performance difference with what's on the PR vs the existing workflow..Looking at ways to speed this up..Will push updates when I get the performance sorted.

File: app/lib/katello/repo_discovery.rb
Comment (sjha4) [reply app/lib/katello/repo_discovery.rb]: This was junk code I was playing with when testing with a specific repo..Removed all of this..

File: app/lib/katello/repo_discovery.rb
Comment (evgeni) [inline app/lib/katello/repo_discovery.rb:13]: The indentation seems off here

File: katello.gemspec
Comment (evgeni) [inline katello.gemspec:51]: Can we drop anemone and webrick deps now?

Comment (sjha4) [reply katello.gemspec]: Removed anemone.. 🎉

File: app/lib/katello/repo_discovery.rb
Comment (ekohl) [inline app/lib/katello/repo_discovery.rb:173]: I think this is a bad idea. The URL object is way more valuable and I'd only add it to `@to_follow` with `to_s`.

File: app/lib/katello/repo_discovery.rb
Comment (ekohl) [inline app/lib/katello/repo_discovery.rb:202]: Looks like the `path` is really the `url`. Please change the name, but better: pass it as a `URI` object and then compare `uri.hostname` to `@uri.hostname`. Also note that hostnames are case insensitive and plain string comparisons are wrong.

File: app/lib/katello/repo_discovery.rb
Comment (ekohl) [inline app/lib/katello/repo_discovery.rb:174]: This should only look at `url.path`. And previously it was `ends_with?`. Now it will also find files within. Is that desirable?

File: katello.gemspec
Comment (ekohl) [inline katello.gemspec:50]:
```suggestion
  # Required for repo discovery
```

File: katello.gemspec
Comment (ekohl) [inline katello.gemspec:53]:
```suggestion
  # Pulp dependencies
```

**Review by ekohl — COMMENTED** [review]
Implementation wise I think you should separate the crawler and Docker search into separate classes. Perhaps even the file crawl as well. Right now it's confusing.

File: app/lib/katello/repo_discovery.rb
Comment (sjha4) [reply app/lib/katello/repo_discovery.rb:173]: We don't necessarily use the URL object properties once we have the list of URLs in the page. It's just matching strings after that point..Becomes easier to pass that around and work with it IMO?

Comment (sjha4) [reply app/lib/katello/repo_discovery.rb]: Updated it in latest push to pass URL object.

File: app/lib/katello/resources/discovery/file_discovery.rb
Comment (ekohl) [inline app/lib/katello/resources/discovery/file_discovery.rb:10]: These parameters are unused. Perhaps drop them? If not, then should the constructor be part of the base class?

File: app/lib/katello/resources/discovery/file_discovery.rb
Comment (ekohl) [inline app/lib/katello/resources/discovery/file_discovery.rb:22]: Since there is no recursion in the method, perhaps just make this `run(resume_point)`?

**Review by ekohl — COMMENTED** [review]
I think this heads in the right direction.

Review wise I wonder if it makes sense to split it up: 1 to create the 3 classes and a follow up that replaces anemone with spidr.

File: app/lib/katello/resources/discovery/file_discovery.rb
Comment (sjha4) [reply app/lib/katello/resources/discovery/file_discovery.rb:10]: I will drop these here..
I was trying to get the constructor as part of base class but the foreman task output which is read in chunks doesn't work when inherited for some reason. Had to go with constructors in child classes and the class_for construct.

Comment (sjha4) [reply app/lib/katello/resources/discovery/file_discovery.rb]: Updated to make credentials and search optional additional params for container and yum classes.

File: app/lib/katello/resources/discovery/file_discovery.rb
Comment (sjha4) [reply app/lib/katello/resources/discovery/file_discovery.rb]: Updated to remove this extra method.

File: app/lib/katello/resources/discovery/container_discovery.rb
Comment (ekohl) [inline app/lib/katello/resources/discovery/container_discovery.rb:8]: This will be `nil` if `upstream_credentials_and_search` is overridden by a caller. If you want it to default to `*` you can use `@search = upstream_credentials_and_search.fetch(:search, '*')`.

File: app/lib/katello/resources/discovery/yum_discovery.rb
Comment (ekohl) [inline app/lib/katello/resources/discovery/yum_discovery.rb:11]: I know it was already this way, but isn't this equivalent?
```suggestion
      @upstream_username = upstream_credentials_and_search[:upstream_username].presence
      @upstream_password = upstream_credentials_and_search[:upstream_password].presence
```

File: app/lib/katello/repo_discovery.rb
Comment (ekohl) [inline app/lib/katello/repo_discovery.rb:2]: I think this should live in the yum class

**Review by ekohl — COMMENTED** [review]
Review wise I always prefer to first have a refactor and then an actual code change. Right now if I read the commits then it's the other way around. That makes this whole change harder to follow, but I don't have commit permissions here so I'll let other reviewers weigh in on that aspect.

File: app/lib/katello/repo_discovery.rb
Comment (parthaa) [inline app/lib/katello/repo_discovery.rb:14]: Can we leverage this => https://github.com/Katello/katello/tree/master/lib/katello/repository_types 
Let the content type tell you the discovery class

File: lib/katello/repo_discovery.rb
Comment (ekohl) [inline lib/katello/repo_discovery.rb:1]: You don't need `expand_path`, but I wonder why you even need this `load` in the first place. I think it's very much incompatible with the Zeitwerk autoloader.

File: app/lib/katello/resources/discovery/container_discovery.rb
Comment (ekohl) [inline app/lib/katello/resources/discovery/container_discovery.rb:2]: I think for the autoloader to work this needs to be `Katello::Resources::Discovery::ContainerDiscovery`. Then given the repetition I'd suggest to rename this to `Katello::Resources::Discovery::Container`.

File: app/lib/katello/repo_discovery.rb
Comment (ekohl) [inline app/lib/katello/repo_discovery.rb:7]: I wouldn't assume a content type and make it a mandatory argument
```suggestion
    def self.class_for(content_type)
```

File: app/lib/katello/repo_discovery.rb
Comment (ekohl) [inline app/lib/katello/repo_discovery.rb:9]: is `_` a method call? If so, I'd prefer braces because it's very easy to miss that it's a method call
```suggestion
      fail _("Content type does not support repo discovery") unless repo_discovery_class
```

File: app/lib/katello/resources/discovery/container_discovery.rb
Comment (ekohl) [inline app/lib/katello/resources/discovery/container_discovery.rb:12]: Like elsewhere, you can use `.presence`
```suggestion
      @upstream_username = upstream_credentials_and_search[:upstream_username].presence
      @upstream_password = upstream_credentials_and_search[:upstream_password].presence
```

File: app/lib/katello/resources/discovery/container_discovery.rb
Comment (ekohl) [inline app/lib/katello/resources/discovery/container_discovery.rb:99]: It's already `nil` or a non-empty string (constructor guarantees that) so why not just:
```suggestion
      request_params[:user] = @upstream_username if @upstream_username
      request_params[:password] = @upstream_password if @upstream_password
```

File: test/lib/repo_discovery_test.rb
Comment (ekohl) [inline test/lib/repo_discovery_test.rb:66]: I'd just hardcode this because it's a better guarantee. If something modifies the hash in place you can silently end up with `q=` instead of `q=busybox`. Previously with a string that could also happen, but it was less likely.
```suggestion
        .with({ method: :get, url: base_url.to_s + "v1/search?q=busybox", headers: {:accept => :json} })
```

**Review by ekohl — CHANGES_REQUESTED** [review]

File: app/lib/katello/resources/discovery/container_discovery.rb
Comment (sjha4) [reply app/lib/katello/resources/discovery/container_discovery.rb]: Aah..Ya..That works..The earlier manual loading was cause the file wasn't getting loaded..Removed that and renamed this..

**Comment by sjha4** [comment]
@ekohl : Updated the PR to address reviews..

File: lib/katello/repo_discovery.rb
Comment (sjha4) [reply lib/katello/repo_discovery.rb]: That was because of what you said in the review below around naming the modules correctly. It loads now without this explicit loading.

**Comment by evgeni** [comment]
@sjha4 will you tackle the packaging side of this, or shall I?

**Comment by sjha4** [comment]
@evgeni : I will give it a try and you can rescue me from time to time..Good learning opportunity cause this is not an urgent thing..

**Comment by evgeni** [comment]
Sounds good to me!

**Comment by evgeni** [comment]
(just don't merge this PR until the packaging is done, otherwise it'll block nightlies)

**Review by parthaa — APPROVED** [review]
Tested with

- Http
- Http With basic auth
- Container images
- local katello instance

Worked great!. ACK


image

https://en.wikipedia.org/wiki/Spyder_(film)

**Comment by sjha4** [comment]
Will rebase and push for tests and merge once packaging is in..

**Review by ekohl — APPROVED** [review]
Packaging will fail until this is merged. Technically you can go through the complex dance of making it work, but then you need to revert all those changes again when you merge it. I'd suggest to merge this.

**Pull request merged**
